### DISCUSSION TITLE
How to get the local or remote address in a handler?

### DISCUSSION THREAD

**Discussion by arniu** [tokio-rs/axum#1376] [Q&A]
### Motivation

We can extract the remote address with `ConnectInfo<SocketAddr>` in  a handler,
while there is no way to get the remote address in it.

### Proposal

Add a new extractor `AddrPair`:

```rust
pub struct AddrPair {
    pub local: SocketAddr,
    pub remote: SocketAddr,
}
```

### Alternatives

Just add `LocalAddr` to get the local address.

**Comment by davidpdrsn** [comment]
We have this already https://docs.rs/axum/0.6.0-rc.1/axum/extract/struct.ConnectInfo.html. Did not find it in the docs or does it not work for you?

**Reply by davidpdrsn** [reply]
And you can get the local addr from the server with https://docs.rs/hyper/latest/hyper/server/struct.Server.html#method.local_addr

**Reply by arniu** [reply]
And how to get the local addr in the handler?

**Reply by davidpdrsn** [reply]
Like any other extractor: https://docs.rs/axum/0.6.0-rc.1/axum/struct.Router.html#method.into_make_service_with_connect_info

**Reply by arniu** [reply]
It seems that I have to write a `MyConnectInfo` to get the local addr.

**Reply by davidpdrsn** [reply]
Does getting it from the server not work?

**Reply by arniu** [reply]
I don't know how to get it from the server...

**Reply by davidpdrsn** [reply]
With the method I linked here https://github.com/tokio-rs/axum/discussions/1376#discussioncomment-3654042

**Reply by arniu** [reply]
And how to get the server in a handler?

**Reply by davidpdrsn** [reply]
Pass it either via `State` or `Extension`

**Reply by arniu** [reply]
Pass the server to the consumed router?

**Reply by davidpdrsn** [reply]
Sorry I forgot that you have to pass the router to the server. You can use `AddrIncoming` from hyper:

```rust
use axum::{
    extract::{ConnectInfo, State},
    routing::get,
    Router,
};
use hyper::server::conn::AddrIncoming;
use std::net::SocketAddr;

#[tokio::main]
async fn main() {
    let addr = SocketAddr::from(([127, 0, 0, 1], 3000));
    let incoming = AddrIncoming::bind(&addr).unwrap();
    let local_addr = incoming.local_addr();

    let app = Router::with_state(AppState { local_addr }).route("/", get(handler));

    axum::Server::builder(incoming)
        .serve(app.into_make_service_with_connect_info::<SocketAddr>())
        .await
        .unwrap();
}

#[derive(Clone)]
struct AppState {
    local_addr: SocketAddr,
}

async fn handler(state: State<AppState>, ConnectInfo(remote_addr): ConnectInfo<SocketAddr>) {
    dbg!(state.local_addr);
    dbg!(remote_addr);
}
```

**Reply by arniu** [reply]
`AddrIncoming::local_addr` differs slightly from `AddrStream::local_addr`.
